modGui/src/org/aion/wallet/crypto/MasterKey.java

```diff
@@ -0,0 +1,45 @@
+package org.aion.wallet.crypto;
+
+import org.aion.crypto.ECKey;
+import org.aion.wallet.exception.ValidationException;
+import org.aion.wallet.util.CryptoUtils;
+
+import java.util.Arrays;
+
+public class MasterKey {
+
+    private final ECKey ecKey;
+
+    public MasterKey(ECKey ecKey) {
+        this.ecKey = ecKey;
+    }
+
+    public ECKey getEcKey() {
+        return ecKey;
+    }
+
+    public ECKey deriveHardened(int[] derivationPath) throws ValidationException {
+        if (derivationPath.length == 0) {
+            throw new ValidationException("Derivation path is incorrect");
+        }
+        byte[] key = ecKey.getPrivKeyBytes();
+        for (final int pathElement : derivationPath) {
+            key = getChild(pathElement, key);
+        }
+        final byte[] seed = Arrays.copyOfRange(key, 0, 32);
+        return new SeededECKeyEd25519(seed);
+
+    }
+
+    private byte[] getChild(final int pathElement, final byte[] keyHash) throws ValidationException {
+        byte[] parentPrivateKey = Arrays.copyOfRange(keyHash, 0, 32);
```

Comment:
should we catch the ArrayIndexOutOfBoundsException exception in case
key length change in the future?

Reply:
Ok